src/disk_io_thread.cpp

```diff
@@ -3174,10 +3139,44 @@ namespace libtorrent
 			add_completed_jobs(completed_jobs);
 	}
 
-	void disk_io_thread::thread_fun(int thread_id, thread_type_t type
-		, boost::shared_ptr<io_service::work> w)
+	bool disk_io_thread::wait_for_job(job_queue& jobq, disk_io_thread_pool& threads
+		, std::unique_lock<std::mutex>& l)
 	{
-		DLOG("started disk thread %d\n", int(thread_id));
+		TORRENT_ASSERT(l.owns_lock());
+		while (jobq.m_queued_jobs.empty())
+		{
+			// if the number of wanted threads is decreased,
+			// we may stop this thread
+			// when we're terminating the last thread, make sure
+			// we finish up all queued jobs first
+			if (threads.should_exit()
+				&& (jobq.m_queued_jobs.empty()
+					|| threads.num_threads() > 1)
```

Comment:
`num_threads()` gets decremented in `try_thread_exit()` so since we hold the job queue mutex for the entire operation it is atomic with respect to the other workers.